### PR TITLE
Add Google Cloud provider

### PR DESCRIPTION
**Pull request by themue**
**What this PR does / why we need it**:

Supporting of Google Cloud with the machine provider.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #

**Special notes for your reviewer**:

```release-note
Add support of Google Cloud as provider
```

### CODE REVIEW

File: pkg/cloudprovider/provider/gce/cloudconfig.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
+//
+// Google Cloud Provider for the Machine Controller
+//
+
+package gce
+
+//-----
+// Imports
+//-----
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/gce/cloudconfig.go]: I'm not a big fan of such comments.
For me this is just noise and decreases readability as `import` keyword is expressive enough for me.
This applies for all "segment" comments in this PR - i'm just commenting on a single bit to have a single conversation.

But this is just my personal opinion.
@alvaroaleman Any opinion on this?

File: pkg/cloudprovider/provider/gce/cloudconfig.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/Masterminds/sprig"
+)
+
+//-----
+// Constants
+//-----
+
+// cloudConfigTemplate renders the cloud-config in gcfg format. All
+// fields are optional, that's why containing the ifs and the explicit newlines.
+// TODO(mue) Check for mandatory fields and better ways to drop empty fields,
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/gce/cloudconfig.go]: AFAIK there is no field which is mandatory

File: pkg/cloudprovider/provider/gce/cloudconfig_test.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	}{
+		{
+			name: "minimum test",
+			config: &cloudConfig{
+				Global: global{
+					ProjectID: "my-project-id",
+					LocalZone: "my-zone",
+				},
+			},
+			contents: "[global]\n" +
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/gce/cloudconfig_test.go]: Please use a multi line string using backticks. It improves readability

File: pkg/cloudprovider/provider/gce/service.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+		return svc.RegionOperations.Get(projectID, region, op.Name).Do()
+	default:
+		return svc.GlobalOperations.Get(projectID, op.Name).Do()
+	}
+}
+
+// waitOperation waits for a GCE operation to be completed or timed out.
+func (svc *service) waitOperation(projectID string, op *compute.Operation, timeout time.Duration) (err error) {
+	started := time.Now()
+	waiting := 100 * time.Millisecond
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/gce/service.go]: Instead if building you own polling function i would recommend to use something which already exists:
https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go#L306

File: pkg/cloudprovider/provider/gce/service.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	"google.golang.org/api/compute/v1"
+)
+
+//-----
+// Constants
+//-----
+
+const (
+	// Timeouts.
+	timeoutNormal = 10 * time.Minute
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/gce/service.go]: Having a global timeout of 10 minutes does not really look right for all API calls we do

File: pkg/cloudprovider/provider/gce/provider.go [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+func New(configVarResolver *providerconfig.ConfigVarResolver) *Provider {
+	return &Provider{
+		resolver: configVarResolver,
+	}
+}
+
+// AddDefaults reads the MachineSpec and applies defaults for provider specific fields
+func (p *Provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, error) {
+	// TODO(mue) Check for defaults to apply.
+	return spec, nil
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/gce/provider.go]: Set:
- disk to `pd-standard`
- diskSize to `25`

File: pkg/cloudprovider/provider/gce/cloudconfig.go (hunk 2 above)
Comment (themue) [reply pkg/cloudprovider/provider/gce/cloudconfig.go]: Will reduce it down to the currently used ones. Also helps to drop these `if` statements in the template.

File: pkg/cloudprovider/provider/gce/cloudconfig_test.go (hunk 3 above)
Comment (themue) [reply pkg/cloudprovider/provider/gce/cloudconfig_test.go]: Got a different opinion when the rest of the construct, like here, is nested and indented. But it's okay.

Comment (mrIncompetent) [reply pkg/cloudprovider/provider/gce/cloudconfig_test.go]: Good point, feel free to keep it like it is :)

File: pkg/cloudprovider/provider/gce/service.go (hunk 4 above)
Comment (themue) [reply pkg/cloudprovider/provider/gce/service.go]: Nice, did not know. Mine has been only stolen. 😄

File: pkg/cloudprovider/provider/gce/service.go (hunk 5 above)
Comment (themue) [reply pkg/cloudprovider/provider/gce/service.go]: Changed to interval of 5 seconds and timeout of 5 minutes as seen in other project.

File: pkg/cloudprovider/provider/gce/cloudconfig.go (hunk 1 above)
Comment (themue) [reply pkg/cloudprovider/provider/gce/cloudconfig.go]: Done.

**Comment by themue** [comment]
> Should the open TODO's in the code be handled later?
> If so, please create an issue & communicate it with @kdomanski

TODOs are done now.

File: examples/gce-machinedeployment.yaml [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            zone: "europe-west3-a"
+            # See https://cloud.google.com/compute/docs/machine-types
+            machineType: "n1-standard-2"
+            # In GB
+            diskSize: 50
+            # Can be 'pd-standard' or 'pd-ssd'
+            diskType: "pd-standard"
+            labels:
+              "kubernetes_cluster": "my-cluster"
+          # Can be 'ubuntu', 'coreos' or 'centos'
```
Comment (mrIncompetent) [inline examples/gce-machinedeployment.yaml]: Not correct anymore. CentOS is not supported

File: pkg/cloudprovider/provider/gce/provider.go [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+		Disks:             disks,
+		Labels:            labels,
+		ServiceAccounts: []*compute.ServiceAccount{
+			{
+				Email: cfg.jwtConfig.Email,
+				Scopes: []string{
+					compute.ComputeScope,
+					compute.DevstorageReadOnlyScope,
+					"https://www.googleapis.com/auth/monitoring",
+					"https://www.googleapis.com/auth/logging.write",
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/gce/provider.go]: Are there any consts for those?

File: examples/gce-machinedeployment.yaml (hunk 7 above)
Comment (themue) [reply examples/gce-machinedeployment.yaml]: Gnah, got me.

File: pkg/cloudprovider/provider/gce/provider.go (hunk 8 above)
Comment (themue) [reply pkg/cloudprovider/provider/gce/provider.go]: Haven't found anywhere, not in `compute` nor elsewhere. But will look again.

Comment (mrIncompetent) [reply pkg/cloudprovider/provider/gce/provider.go]: Maybe you want to temporarily disable the vendor pruning, that helps sometimes finding such things

Comment (themue) [reply pkg/cloudprovider/provider/gce/provider.go]: Looks bad, only those few defined here: https://godoc.org/google.golang.org/api/compute/v1#pkg-constants.

Comment (themue) [reply pkg/cloudprovider/provider/gce/provider.go]: Found packages for logging with scope constants and `cloud.google.com/go/monitoring/apiv3`. The latter is only exporting `DefaultAuthScopes()` returning a slice of all scopes the package needs.

**Pull request merged**